### PR TITLE
KAFKA-14209 : Change Topology optimization to accept list of rules 1/3

### PR DESCRIPTION
**Pull request by vpapavas**
This PR is part of a series implementing the self-join rewriting. As part of it, we decided to clean up the `TOPOLOGY_OPTIMIZATION_CONFIG` and make it a list of optimization rules.  Acceptable values are: NO_OPTIMIZATION, OPTIMIZE which applies all optimization rules or a comma separated list of specific optimizations.

Added unit tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -233,6 +235,15 @@ public class StreamsConfig extends AbstractConfig {
...
     public static final String CLIENT_TAG_PREFIX = "client.tag.";
 
+    /** {@code topology.optimization} */
+    public static final String TOPOLOGY_OPTIMIZATION_CONFIG = "topology.optimization";
+    private static final String TOPOLOGY_OPTIMIZATION_DOC = "A configuration telling Kafka "
+        + "Streams if it should optimize the topology and what optimizations to apply. "
+        + "Acceptable values are: \"+NO_OPTIMIZATION+\", \"+OPTIMIZE+\", "
+        + "or a comma separated list of specific optimizations: "
+        + "(\"+REUSE_KTABLE_SOURCE_TOPICS+\", \"+MERGE_REPARTITION_TOPICS+\"). "
+        + "Disabled by default.";
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]:
```suggestion
        + "\"" + NO_OPTIMIZATION+"\" by default.";
```

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 2]
```diff
@@ -1653,6 +1678,34 @@ private Map<String, Object> getClientCustomProps() {
...
+        final Set<String> verifiedConfigs = new HashSet<>();
+        // Verify it doesn't contain none or all plus a list of optimizations
+        if (configs.contains(NO_OPTIMIZATION) || configs.contains(OPTIMIZE)) {
+            if (configs.size() > 1) {
+                throw new ConfigException("A topology can either not be optimized with " + NO_OPTIMIZATION + " "
+                                              + "or optimized. If you want to optimize the "
+                                              + "topology, you can choose between all "
+                                              + "optimizations with " + OPTIMIZE + " " + "or "
+                                              + "specific optimizations by specifying a comma "
+                                              + "separated list.");
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I took a closer look at the text of this error, and I'm afraid it might be confusing to users. The rendered text will say "A topology can either not be optimized with none or optimized.", which seems hard to parse to me, both because the sentence structure is confusing and because the config to turn off optimizations is mentioned ("none"), but the config to turn them on is not ("all").

```suggestion
                throw new ConfigException(
                    "\"" + config + "\" is not a valid optimization config. " +
                    "Valid optimization configs are: " +
                    "\"" + OPTIMIZE + "\" to enable all optimizations, " +
                    "\"" + NO_OPTIMIZATION + "\" to disable all optimizations, " +
                    "or a comma separated list of specific optimizations: " +
                    "(\"" + REUSE_KTABLE_SOURCE_TOPICS + "\", " +
                    MERGE_REPARTITION_TOPICS + "\"). "
                );
```

Realistically, I'd probably build the string for the second part of that message ("Valid ...") once and save it in a private constant for use in both the doc string as well as the validation error.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 3]
```diff
@@ -1263,6 +1263,54 @@ public void testInvalidSecurityProtocol() {
...
+    @Test
+    public void shouldThrowExceptionWhenTopologyOptimizationDoesNotExist() {
+        final String value = String.join(",", "topology.optimization.does.not.exist");
+        props.put(TOPOLOGY_OPTIMIZATION_CONFIG, value);
+        final ConfigException exception = assertThrows(ConfigException.class, () -> new StreamsConfig(props));
+        assertTrue(exception.getMessage().contains("Unrecognized config."));
+    }
+
+    @Test
+    public void shouldAllowMultipleOptimizations() {
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:1303]: Can we add a test that "all" results in all the expected optimizations being listed? And that "none" results in an empty list?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 4]
```diff
@@ -243,6 +254,22 @@ public class StreamsConfig extends AbstractConfig {
      */
     public static final String OPTIMIZE = "all";
 
+    /**
+     * Config value for parameter {@link #TOPOLOGY_OPTIMIZATION_CONFIG "topology.optimization"}
+     * for enabling the specific optimization that reuses source topic as changelog topic
+     * for KTables.
+     */
+    public static final String REUSE_KTABLE_SOURCE_TOPICS = "topology.optimization.reuse.ktable.source.topics";
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: In the KIP, seems we do not have the `topology.optimization.` prefix for such values? Also I personally feel that we do not need those prefixes since for users who do want to list specific rules, they would become too long.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 5]
```diff
@@ -1263,6 +1263,54 @@ public void testInvalidSecurityProtocol() {
...
+    @Test
+    public void shouldThrowExceptionWhenTopologyOptimizationOffAndSet() {
+        final String value = String.join(",", StreamsConfig.NO_OPTIMIZATION, StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS);
+        props.put(TOPOLOGY_OPTIMIZATION_CONFIG, value);
+        final ConfigException exception = assertThrows(ConfigException.class, () -> new StreamsConfig(props));
+        assertTrue(exception.getMessage().contains("A topology can either not be optimized with"));
+    }
+
+    @Test
+    public void shouldThrowExceptionWhenOptimizationDoesNotExist() {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: What's the main difference between `shouldThrowExceptionWhenOptimizationDoesNotExist` and `shouldThrowExceptionWhenTopologyOptimizationDoesNotExist`? Could they be just one test?

**Review by guozhangwang — COMMENTED** [review]
Could we also add to the docs files (`upgrade-guide.html`) about this config value semantics change? Other comments are pretty minor, thanks!

**Review by vvcephei — COMMENTED** [review]
Thanks, @vpapavas ! Just a few small pieces of feedback.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java (hunk 5 above)
Comment (vpapavas) [reply streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java]: The difference is that the first has a bad config in a list of valid configs and the other has only the bad config. I changegd the name to make it clearer.

File: docs/streams/developer-guide/config-streams.html [hunk 6]
```diff
@@ -424,8 +424,8 @@ <h4><a class="toc-backref" href="#id23">num.standby.replicas</a><a class="header
           </tr>
           <tr class="row-even"><td>topology.optimization</td>
             <td>Medium</td>
-            <td colspan="2">A configuration telling Kafka Streams if it should optimize the topology</td>
-            <td>none</td>
+            <td colspan="2">A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>NO_OPTIMIZATION</code>, <code>OPTIMIZE</code> or a comma separated list of specific optimizations: (<code>REUSE_KTABLE_SOURCE_TOPICS</code>, <code>MERGE_REPARTITION_TOPICS</code>). </td>
```
Comment (vvcephei) [inline docs/streams/developer-guide/config-streams.html]:
```suggestion
            <td colspan="2">A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: (<code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>)). </td>
```

File: docs/streams/developer-guide/config-streams.html [hunk 7]
```diff
@@ -942,7 +942,7 @@ <h4><a class="toc-backref" href="#id31">topology.optimization</a><a class="heade
       <blockquote>
         <div>
           <p>
-            You can tell Streams to apply topology optimizations by setting this config. The optimizations are currently all or none and disabled by default.
+            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>NO_OPTIMIZATION</code>, <code>OPTIMIZE</code> or a comma separated list of specific optimizations: (<code>REUSE_KTABLE_SOURCE_TOPICS</code>, <code>MERGE_REPARTITION_TOPICS</code>).
```
Comment (vvcephei) [inline docs/streams/developer-guide/config-streams.html]:
```suggestion
            A configuration telling Kafka Streams if it should optimize the topology and what optimizations to apply. Acceptable values are: <code>StreamsConfig.NO_OPTIMIZATION</code> (<code>none</code>), <code>StreamsConfig.OPTIMIZE</code> (<code>all</code>) or a comma separated list of specific optimizations: (<code>StreamsConfig.REUSE_KTABLE_SOURCE_TOPICS</code> (<code>reuse.ktable.source.topics</code>), <code>StreamsConfig.MERGE_REPARTITION_TOPICS</code> (<code>merge.repartition.topics</code>)).
          </p>
          <p>
            We recommend listing specific optimizations in the config for production code so that the structure of your topology will not change unexpectedly during upgrades of the Streams library.
          </p>
          <p>
```

File: docs/streams/developer-guide/config-streams.html [hunk 8]
```diff
@@ -942,7 +942,7 @@ <h4><a class="toc-backref" href="#id31">topology.optimization</a><a class="heade
...
             These optimizations include moving/reducing repartition topics and reusing the source topic as the changelog for source KTables. It is recommended to enable this.
           </p>
```
Comment (vvcephei) [inline docs/streams/developer-guide/config-streams.html]:
```suggestion
            These optimizations include moving/reducing repartition topics and reusing the source topic as the changelog for source KTables. These optimizations will save on network traffic and storage in Kafka without changing the semantics of your applications. Enabling them is recommended.
          </p>
```

**Review by vvcephei — APPROVED** [review]
Awesome! Thanks, @vpapavas .

I'm proposing a slight modification to the docs. I'll go ahead and merge them and then merge this PR when the tests pass.

**Comment by vvcephei** [comment]
That red `x` is spurious. The build is ongoing: https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-12641/6/

**Comment by guozhangwang** [comment]
LGTM. Thanks @vpapavas
